report_scale_down_metric.py

```diff
@@ -0,0 +1,216 @@
+            instance_count += get_current_cluster_size(cluster_name, next_token=lci_result['nextToken'], max_results=max_results)
+        else:
+            if 'containerInstanceArns' in lci_result:
+                instance_count = len(lci_result['containerInstanceArns'])
+        return instance_count
+
+
+    def get_stack_parameters(stack_name):
+        '''Get the thresholds for CPU and Memory for scaling down from the stack'''
+        ds_result = cloudformation.describe_stacks(StackName=stack_name)
+        if 'Stacks' in ds_result:
+            if 'Parameters' in ds_result['Stacks'][0]:
+                return ds_result['Stacks'][0]['Parameters']
+
+    min_cluster_size = 1
+    if stack_name:
+        stack_params = get_stack_parameters(stack_name)
+        for param in stack_params:
+            if param['ParameterKey'] == SCALE_DOWN_CPU_RESERVATION:
+                cpu_threshold = int(param['ParameterValue'])
+            elif param['ParameterKey'] == SCALE_DOWN_MEM_RESERVATION:
+                mem_threshold = int(param['ParameterValue'])
+            elif param['ParameterKey'] == CLUSTER_MIN_SIZE:
+                min_cluster_size = int(param['ParameterValue'])
+
+    if not cpu_threshold or not mem_threshold:
+        logger.critical('Not able to determine scale down CPU or Memory thresholds - aborting')
+        exit(1)
+
+    # Get end_time - now
+    # Get start_time - 5 minutes ago
+    end_time = datetime.utcnow()
+    end_time_utc = end_time.isoformat()
+    start_time = end_time - timedelta(minutes=5)
+    start_time_utc = start_time.isoformat()
+
+    avg_stats = get_cluster_cpu_and_mem_reservation(cluster_name, start_time_utc, end_time_utc)
+
+    scale_down_cpu = False
+    if 'CPU' in avg_stats:
+        if int(avg_stats['CPU']) < cpu_threshold:
+            logger.debug('Based on CPU, need a scale down')
+            scale_down_cpu = True
+        else:
+            logger.debug('Based on CPU, DO NOT need a scale down')
+
+    scale_down_mem = False
+    if 'Mem' in avg_stats:
+        if int(avg_stats['Mem']) < mem_threshold:
+            logger.debug('Based on Memory, need a scale down')
+            scale_down_mem = True
+        else:
+            logger.debug('Based on Memory, DO NOT need a scale down')
+
+    current_cluster_size = get_current_cluster_size(cluster_name)
+
+    scale_down_metric = 0
+    if scale_down_cpu and scale_down_mem:
+        if current_cluster_size > min_cluster_size:
+            logger.debug('Both CPU and Memory are below thresholds, and cluster size is above Min Cluster Size - post a scale down metric')
+            scale_down_metric = 1
+        else:
+            logger.debug('Both CPU and Memory are below thresholds, but cluster size is already at Min Cluster Size')
+
+    #Report scale down metric to CloudWatch
+    put_cloudwatch_metric(scale_down_metric)
+
+
+if __name__ == "__main__":
+
+    LOG_FILENAME = '/var/log/ecs/report_scale_down_metric.log'
```

Comment:
Do we send this to papertrail?  If not, we should because otherwise it will be hard to troubleshoot scaling events here (I'm thinking of a case where 2 instances possibly send different values for the scale down metric)